### PR TITLE
adding preview attached image & upload from clipboard

### PR DESCRIPTION
**Pull request by floatinghotpot**
Added 2 features:

1. Previously, the attached image thumbnail is too small to view. Now clicking the thumbnail will display a popup to view the full-size image. If the image is larger than 300px, the popup will also be centered.

2. Previously, only support uploading from file system. Now also support uploading image from clipboard. Just press Ctrl-V to paste to preview, then click "clipboard", a image with name "clipboard.png" will be uploaded.

### CODE REVIEW

File: client/components/cards/attachments.jade [hunk 1]
```diff
@@ -3,6 +3,13 @@ template(name="cardAttachmentsPopup")
     li
       input.js-attach-file.hide(type="file" name="file" multiple)
       a.js-computer-upload {{_ 'computer'}}
+    li
+      a.js-upload-clipboard-image {{_ 'clipboard'}}
```
Comment (mquandalle) [inline client/components/cards/attachments.jade]: You should add the corresponding key in the `i18n/en.i18n.json` file (sorted by alphabetical order).

File: client/components/cards/attachments.jade [hunk 2]
```diff
@@ -3,6 +3,13 @@ template(name="cardAttachmentsPopup")
...
+      img.preview-clipboard-image(src="")
+
+template(name="previewAttachedImagePopup")
+  p {{_ "preview"}}
```
Comment (mquandalle) [inline client/components/cards/attachments.jade]: Same here, add the i18n key.

File: client/components/cards/attachments.js [hunk 3]
```diff
@@ -1,5 +1,12 @@
 Template.attachmentsGalery.events({
-  'click .js-add-attachment': Popup.open('cardAttachments'),
+  'click .js-add-attachment'() {
+    Popup.open('cardAttachments').apply(this, arguments);
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: This should use ES6:

```es6
  'click .js-add-attachment'(...args) {
    Popup.open('cardAttachments').call(this, ...args);
```

File: client/components/cards/attachments.js [hunk 4]
```diff
@@ -1,5 +1,12 @@
...
+
+    $('html').pasteImageReader(function(results) {
+      Popup.pasted = results;
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: I’m a bit confused about this `Popup.pasted`, could you add a comment on it

File: client/components/cards/attachments.js [hunk 5]
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
...
+      Attachments.insert(file);
+    }
+
+    $('html').pasteImageReader(function(){});
+    Popup.close();
+  },
 });
+
+// ------------------------------------------------------------------------
+// Created by STRd6
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: You should move this code in a dedicated file, also I would recommend to keep it unmodified. Adding

```js
/*eslint-disable */
```

at the beginning of the file should disabling Eslint for it.

File: client/components/cards/attachments.js [hunk 6]
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
   'click .js-remove-cover'() {
     Cards.findOne(this.cardId).unsetCover();
   },
+  'click .js-preview-image'() {
+    Popup.open('previewAttachedImage').apply(this, arguments);
+
+    setTimeout(function(){
+      const img = $('img.preview-large-image');
+      const w = img.width(), h = img.height();
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Please define a single variable per line (repeat `const`).

File: client/components/cards/attachments.js [hunk 7]
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
...
+      if(w > 300) {
+        $('div.pop-over').css({
+          width: (w + 20),
+          position: 'absolute',
+          left: (window.innerWidth - w)/2,
+          top: (window.innerHeight - h)/2,
+        });
+      }
+    }, 50);
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Could you explain the reason for this delay — and is there any chance we could remove it at some point?

File: client/components/cards/attachments.js [hunk 8]
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
   'click .js-remove-cover'() {
     Cards.findOne(this.cardId).unsetCover();
   },
+  'click .js-preview-image'() {
+    Popup.open('previewAttachedImage').apply(this, arguments);
+
+    setTimeout(function(){
+      const img = $('img.preview-large-image');
+      const w = img.width(), h = img.height();
+      if(w > 300) {
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: We put a space after `if`, I’ll see how to add that in the linting rules.

File: client/components/cards/attachments.js [hunk 9]
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
     tpl.find('.js-attach-file').click();
     evt.preventDefault();
   },
+  'click .js-upload-clipboard-image'() {
+    const results = Popup.pasted;
+    if((typeof Blob !== 'undefined') && (results.file instanceof Blob) && results.file) {
+      const card = this;
+      const file = new FS.File( results.file );
+      if(typeof results.file.type === 'string') {
+        const extname = results.file.type.split('/')[1];
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: IIRC CollectionFS has a method to extract a file extension name.

File: client/components/cards/attachments.js [hunk 10]
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
...
+        const filename = 'clipboard.png'.replace('png', extname);
+        file.name( filename );
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Style: no space inside the parents

File: client/components/cards/attachments.js [hunk 11]
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
...
+      }
+      file.updatedAt(new Date());
+      file.boardId = card.boardId;
+      file.cardId  = card._id;
+      Attachments.insert(file);
+    }
+
+    $('html').pasteImageReader(function(){});
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Could you add a comment about what this does — I guess, reset?

File: client/components/cards/attachments.js [hunk 12]
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
...
+      const card = this;
+      const file = new FS.File( results.file );
+      if(typeof results.file.type === 'string') {
+        const extname = results.file.type.split('/')[1];
+        const filename = 'clipboard.png'.replace('png', extname);
+        file.name( filename );
+      }
+      file.updatedAt(new Date());
+      file.boardId = card.boardId;
+      file.cardId  = card._id;
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Style: don’t align `=`

**Comment by mquandalle** [comment]
Hi @floatinghotpot,

Thank you for this PR, it looks great and I’ll be happy to merge it. I’ve written some relatively easy to fix (or answer) inline comments, and I’ll appreciate if you could take a look at them.

I also think that the preview should be on a separate popup displayed after the click on “Clipboard” (`Popup.open("clipboardPreview")`) and have a small text that explain what it does — “The following image is available in you clipboard, do you whish to attach it to this card” (or any better wording really).

**Comment by mquandalle** [comment]
And we should also handle the case when there is no image the clipboard correctly — currently it only displays a broken image.

**Comment by floatinghotpot** [comment]
Hi @mquandalle, thanks for the precise comments. I like the style. Will improve accordingly.

File: client/components/cards/attachments.js (hunk 11 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: Yes, clean up the Ctrl-V handler, as only allow it working in preview popup.

File: client/components/cards/attachments.js (hunk 9 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: results.file.type is usually "image/png", not a filename.

File: client/components/cards/attachments.js (hunk 7 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: When there are multiple thumbnail, if click one then another very fast, will get a wrong width value. Adding a little delay will avoid the bug.

File: client/components/cards/attachments.js (hunk 5 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: Shall I put the separated code under /lib? where to put the link to it please?

File: client/components/cards/attachments.js (hunk 4 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: The paste can only be fired by keyboard Ctrl-V. Here we temporarily store the content in Popup.pasted, later we get from it then upload.

File: client/components/cards/attachments.js (hunk 3 above)
Comment (floatinghotpot) [inline client/components/cards/attachments.js]: Good advice, I will try and verify. 

BTW, some ES6 rules seems not working in my Chrome, like string template, weird.

**Comment by floatinghotpot** [comment]
Screenshot:
---
Menu:
menu
---
Preview:
preview

**Comment by floatinghotpot** [comment]
Hi @mquandalle, I've fixed according to your comments. 
And, also add 2 more: 
1. not only allow paste image from clipboard, but also drag & drop image file.
2. translated zh-CN.

**Comment by mquandalle** [comment]
Thanks for improving your PR, I’ll do a second review pass. However I’m not managing non-English translations via GitHub PRs (that would be difficult) but instead we use Transifex. So I encourage you to upload your Chinese file to Transifex (they may be some updates there since the last time I pulled the translations, I do that before each release), and could you please remove them from these PR (you can `git --amend`, `git push -f`).

File: client/components/cards/attachments.js [hunk 13]
```diff
@@ -20,6 +20,33 @@ Template.attachmentsGalery.events({
   'click .js-remove-cover'() {
     Cards.findOne(this.cardId).unsetCover();
   },
+  'click .js-preview-image': Popup.open('previewAttachedImage'),
+});
+
+Template.previewAttachedImagePopup.onRendered(function() {
+  // when there are multiple thumbnails,
+  // if click one then another very fast, we will get a wrong width value.
+  // adding a little delay will avoid the bug.
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Could you try removing the `setTimeout` and executing `Tracker.flush()` before measuring the image to see if it works in the edge case you are mentioning?

File: client/components/cards/attachments.js [hunk 14]
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
     tpl.find('.js-attach-file').click();
     evt.preventDefault();
   },
+  'click .js-upload-clipboard-image': Popup.open('previewClipboardImage'),
+});
+
+Template.previewClipboardImagePopup.onRendered(function() {
+  // we can paste image from clipboard
+  $('html').pasteImageReader(function(results) {
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: I think `$(document.body)` is a more accepted convention, right?

File: client/lib/dropImage.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+            };
+            reader.readAsDataURL(f);
+            return;
+          }
+        }
+      });
+    });
+  };
+})(jQuery);
+
```
Comment (mquandalle) [inline client/lib/dropImage.js]: Please remove the extra line

File: client/lib/pasteImage.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+            };
+            reader.readAsDataURL(f);
+            return;
+          }
+        }
+      });
+    });
+  };
+})(jQuery);
+
```
Comment (mquandalle) [inline client/lib/pasteImage.js]: same

File: client/components/cards/attachments.js [hunk 17]
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
...
+    }
+  });
+});
+
+Template.previewClipboardImagePopup.events({
+  'click .js-upload-pasted-image'() {
+    const results = Popup.pasted;
+    if(results && results.file) {
+      const card = this;
+      const file = new FS.File( results.file );
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Style: no space between parents

File: client/components/cards/attachments.js [hunk 18]
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
...
+      if(!results.name) {
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: Style: a space after `if`

File: client/components/cards/attachments.js [hunk 19]
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
...
+    if(results.dataURL.startsWith('data:image/')) {
+      Popup.pasted = results;
+      $('img.preview-clipboard-image').attr('src', results.dataURL);
+    }
+  });
+
+  // we can also drag & drop image file to it
+  $('html').dropImageReader(function(results) {
+    if(results.dataURL.startsWith('data:image/')) {
+      Popup.pasted = results;
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: You should use a local variable and not pollute the Popup object (which is too fragile and will cause bugs). Just create something `let droppedImage;` on top of this file, and use it here.

File: client/components/cards/attachments.js [hunk 20]
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
...
+        if(typeof results.file.type === 'string') {
+          file.name(results.file.type.replace('image/', 'clipboard.'));
+        }
+      }
+      file.updatedAt(new Date());
+      file.boardId = card.boardId;
+      file.cardId = card._id;
+      Attachments.insert(file);
+
+      delete Popup.pasted;
```
Comment (mquandalle) [inline client/components/cards/attachments.js]: to remove

**Comment by floatinghotpot** [comment]
Hi @mquandalle, I've improved the PR again, hope this time can pass your review.

BTW, besides this PR, in my devel branch, I've implemented some features for our production purpose in real project:
* 7c47384 add: import members/labels/lists as template from other board
* 63b9a64 add: import card from other board list
* 869735e allow creating multi cards by pasting excel tsv with field 'title','desc','duedate','member','label'
* 297f5e9 add task date to card, auto record date when drop to list, if status defined for list
* 7f84285 extend card and list model for task date & status
* 4a3525b improve PR, allow paste / drag&drop to upload
* 955fabf add preview attached image & allow upload image from clipboard

I can create separated PR with each feature, are you interested?

**Comment by mquandalle** [comment]
Thanks for improving your PR!

I’m very confused by the broken image by default:

wekan-attachments

Could you remove it if there is no valid image to display. Also I would to instructions on first line (before the image) and add some vertical margin. The button should also have the class name `.primary`.

**Comment by floatinghotpot** [comment]
I removed the default src="", which is white blank in my chrome.

Now changed to:
screen

**Comment by floatinghotpot** [comment]
@mquandalle I've merged the changes into 1 commit. Is there anything else for me to improve in this PR?

**Comment by mquandalle** [comment]
It’s all good, I expect to merge your PR this week-end :)

**Comment by floatinghotpot** [comment]
@mquandalle I am sorry to hear about the terrible things happened in Paris. Hope you are doing well.

**Comment by mquandalle** [comment]
Yes, my family and friends are safe. I’m not currently in France (I’m in San Francisco) but I spent a lot of time watching the news and calling my relatives this week-end and I haven’t worked at all.

**Pull request merged**